ts/Stock/Indicators/Klinger/KlingerIndicator.ts

```diff
@@ -0,0 +1,434 @@
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line. If not set, it's inherited from
+                 * [plotOptions.klinger.color
+                 * ](#plotOptions.sma.color).
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#ff0000'
+            }
+        },
+        dataGrouping: {
+            approximation: 'averages'
+        },
+        tooltip: {
+            pointFormat: '<span style="color: {point.color}">\u25CF</span><b> {series.name}</b><br/>' +
+                '<span style="color: {point.color}">Klinger</span>: {point.y}<br/>' +
+                '<span style="color: {point.series.options.signal.styles.lineColor}">Signal</span>' +
+                    ': {point.signal}<br/>'
+        }
+    } as KlingerOptions);
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    public data: Array<KlingerPoint> = void 0 as any;
+    public points: Array<KlingerPoint> = void 0 as any;
+    public options: KlingerOptions = void 0 as any;
+    public volumeSeries: LineSeries = void 0 as any;
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    init(this: KlingerIndicator): void {
+        const args = arguments,
+            ctx = this;
+
+        // Check if the EMA module is added.
+        RequiredIndicatorMixin.isParentLoaded(
+            (EMAIndicator as any),
+            'ema',
+            ctx.type,
+            function (indicator: Highcharts.Indicator): undefined {
+                indicator.prototype.init.apply(ctx, args);
+                return;
+            }
+        );
+    }
+
+    public calculateTrend(
+        this: KlingerIndicator,
+        yVal: (Array<Array<number>>),
+        i: number
+    ): number {
+        const isUpward = yVal[i][1] + yVal[i][2] + yVal[i][3] >
+            yVal[i - 1][1] + yVal[i - 1][2] + yVal[i - 1][3];
+
+        return isUpward ? 1 : -1;
+    }
+
+    // Checks if the series and volumeSeries are accessible, number of
+    // points.x is longer than period, is series has OHLC data
+    public isValidData(
+        this: KlingerIndicator,
+        firstYVal: Array<number>
+    ): boolean {
+        const chart = this.chart,
+            options: KlingerOptions = this.options,
+            series = this.linkedParent,
+            isSeriesOHLC: boolean = isArray(firstYVal) && firstYVal.length === 4,
+            volumeSeries = (this.volumeSeries ||
+                    (
+                        this.volumeSeries =
+                        chart.get((options.params as any).volumeSeriesID) as any
+                    )
+            );
+
+        /**
+         * @private
+         * @param {Highcharts.Series} series to check length validity on.
+         * @return {boolean|undefined} true if length is valid.
+         */
+        function isLengthValid(series: LineSeries): (boolean|undefined) {
```

Comment:
Is it necessary to define this function on every indicator recalc? If this is a utility method, then move it to the top scope